File: api/app/schemas/topic.py
"""
Topic schemas.
"""
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime
from app.models.enums import TopicVisibility


class TopicResponse(BaseModel):
    """Topic response schema."""
    id: int
    name: str
    description: Optional[str] = None
    icon: Optional[str] = None
    created_by_user_id: int
    visibility: TopicVisibility = TopicVisibility.PRIVATE
    liked: int = 0
    created_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


class CreateTopicRequest(BaseModel):
    """Request schema for creating a topic."""
    name: str
    description: Optional[str] = None
    icon: Optional[str] = None
    user_id: int  # Keep as user_id for backward compatibility in API, maps to created_by_user_id
    visibility: Optional[TopicVisibility] = TopicVisibility.PRIVATE


class UpdateTopicRequest(BaseModel):
    """Request schema for updating a topic."""
    name: Optional[str] = None
    description: Optional[str] = None
    icon: Optional[str] = None
    visibility: Optional[TopicVisibility] = None


class TopicsResponse(BaseModel):
    """Response schema for topics list."""
    topics: List[TopicResponse]

File: app/lib/src/features/shared/domain/topic.dart
class Topic {
  final int id;
  final String name;
  final String? description;
  final String? icon;
  final int createdByUserId;
  final String visibility; // 'public' or 'private'
  final int liked;
  final DateTime? createdAt;

  Topic({
    required this.id,
    required this.name,
    this.description,
    this.icon,
    required this.createdByUserId,
    this.visibility = 'private',
    this.liked = 0,
    this.createdAt,
  });

  factory Topic.fromJson(Map<String, dynamic> json) {
    return Topic(
      id: json['id'] as int,
      name: json['name'] as String,
      description: json['description'] as String?,
      icon: json['icon'] as String?,
      createdByUserId: json['created_by_user_id'] as int,
      visibility: json['visibility'] as String? ?? 'private',
      liked: json['liked'] as int? ?? 0,
      createdAt: json['created_at'] != null 
          ? DateTime.parse(json['created_at'] as String)
          : null,
    );
  }
}

